File: packedgo/back/order-service/src/main/java/com/packed_go/order_service/service/OrderService.java
package com.packed_go.order_service.service;

import java.util.List;

import com.packed_go.order_service.dto.MultiOrderCheckoutResponse;
import com.packed_go.order_service.dto.SessionStateResponse;
import com.packed_go.order_service.dto.request.CheckoutRequest;
import com.packed_go.order_service.dto.request.PaymentCallbackRequest;
import com.packed_go.order_service.dto.response.CheckoutResponse;
import com.packed_go.order_service.entity.MultiOrderSession;
import com.packed_go.order_service.entity.Order;

public interface OrderService {
    
    /**
     * Procesa el checkout del carrito activo del usuario
     */
    CheckoutResponse checkout(Long userId, CheckoutRequest request);
    
    /**
     * Procesa el checkout multitenant creando múltiples órdenes si hay items de varios admins
     */
    MultiOrderCheckoutResponse checkoutMulti(Long userId);
    
    /**
     * Backend State Authority: Obtiene o crea la sesión actual del usuario
     * Busca sesión activa (PENDING/PARTIAL no expirada), si no existe crea nueva desde cart
     * El frontend NO necesita guardar sessionId, solo llamar a este endpoint con JWT
     */
    SessionStateResponse getCurrentCheckoutState(Long userId);
    
    /**
     * Obtiene el estado de una sesión de múltiples órdenes
     */
    MultiOrderCheckoutResponse getSessionStatus(String sessionId);
    
    /**
     * Recupera una sesión usando el token de recuperación (sin requerir JWT)
     */
    MultiOrderCheckoutResponse recoverSessionByToken(String sessionToken);
    
    /**
     * Obtiene todos los tickets generados en una sesión
     */
    List<Object> getSessionTickets(String sessionId);
    
    /**
     * Abandona una sesión de checkout y devuelve items al carrito si no hay pagos completados
     */
    void abandonSession(String sessionId, Long userId);
    
    /**
     * Actualiza el estado de una orden basado en el callback del pago
     */
    void updateOrderFromPaymentCallback(PaymentCallbackRequest request);
    
    /**
     * Obtiene todas las órdenes de un usuario
     */
    List<Order> getUserOrders(Long userId);
    
    /**
     * Obtiene una orden por su número
     */
    Order getOrderByNumber(String orderNumber);
    
    /**
     * Obtiene una orden por ID
     */
    Order getOrderById(Long orderId);
}

File: packedgo/front-angular/src/app/core/services/order.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders, HttpErrorResponse } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { 
  MultiOrderCheckoutResponse, 
  Order, 
  MultiOrderSession 
} from '../../shared/models/order.model';

@Injectable({
  providedIn: 'root'
})
export class OrderService {
  private http = inject(HttpClient);
  private apiUrl = environment.ordersServiceUrl;

  /**
   * Obtiene los headers con el token JWT
   */
  private getHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`
    });
  }

  /**
   * Realiza el checkout multi-admin
   * Agrupa los items del carrito por adminId y crea múltiples órdenes
   * @returns Observable con la respuesta del checkout incluyendo los grupos de pago
   */
  checkoutMulti(): Observable<MultiOrderCheckoutResponse> {
    return this.http.post<MultiOrderCheckoutResponse>(
      `${this.apiUrl}/orders/checkout/multi`,
      {}, // Body vacío, el backend obtiene el userId del token
      { headers: this.getHeaders() }
    ).pipe(
      tap(response => {
        console.log('Checkout multi completado:', response);
      }),
      catchError(this.handleError)
    );
  }

  /**
   * Obtiene el estado de una sesión multi-orden
   * @param sessionId ID de la sesión
   * @returns Observable con el estado actualizado de la sesión
   */
  getSessionStatus(sessionId: string): Observable<MultiOrderCheckoutResponse> {
    return this.http.get<MultiOrderCheckoutResponse>(
      `${this.apiUrl}/orders/sessions/${sessionId}`,
      { headers: this.getHeaders() }
    ).pipe(
      tap(response => {
        console.log('Estado de sesión:', response);
      }),
      catchError(this.handleError)
    );
  }

  /**
   * Recupera una sesión usando el token de recuperación (sin JWT)
   * @param sessionToken Token de recuperación de la sesión
   * @returns Observable con el estado de la sesión
   */
  recoverSession(sessionToken: string): Observable<MultiOrderCheckoutResponse> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
      'X-Session-Token': sessionToken
    });

    return this.http.get<MultiOrderCheckoutResponse>(
      `${this.apiUrl}/orders/session/recover`,
      { headers }
    ).pipe(
      tap(response => {
        console.log('Sesión recuperada:', response);
      }),
      catchError(this.handleError)
    );
  }

  /**
   * Obtiene todos los tickets generados en una sesión
   * @param sessionId ID de la sesión
   * @returns Observable con la lista de tickets
   */
  getSessionTickets(sessionId: string): Observable<any[]> {
    return this.http.get<any[]>(
      `${this.apiUrl}/orders/session/${sessionId}/tickets`
    ).pipe(
      tap(response => {
        console.log('Tickets de sesión:', response);
      }),
      catchError(this.handleError)
    );
  }

  /**
   * Abandona una sesión de checkout y devuelve los items al carrito
   * Solo funciona si no hay pagos completados en la sesión
   * @param sessionId ID de la sesión a abandonar
   * @returns Observable con el resultado de la operación
   */
  abandonSession(sessionId: string): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.apiUrl}/orders/sessions/${sessionId}/abandon`,
      {}, // Body vacío
      { headers: this.getHeaders() }
    ).pipe(
      tap(response => {
        console.log('Sesión abandonada:', response);
      }),
      catchError(this.handleError)
    );
  }

  /**
   * Obtiene todas las órdenes del usuario actual
   * @returns Observable con el array de órdenes
   */
  getUserOrders(): Observable<Order[]> {
    return this.http.get<Order[]>(
      `${this.apiUrl}/orders/user`,
      { headers: this.getHeaders() }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Obtiene una orden específica por ID
   * @param orderId ID de la orden
   * @returns Observable con los detalles de la orden
   */
  getOrderById(orderId: number): Observable<Order> {
    return this.http.get<Order>(
      `${this.apiUrl}/orders/${orderId}`,
      { headers: this.getHeaders() }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Maneja errores HTTP
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    let errorMessage = 'Ocurrió un error desconocido';
    
    if (error.error instanceof ErrorEvent) {
      // Error del lado del cliente
      errorMessage = `Error: ${error.error.message}`;
    } else {
      // Error del lado del servidor
      switch (error.status) {
        case 400:
          errorMessage = error.error?.message || 'Datos inválidos';
          break;
        case 401:
          errorMessage = 'Tu sesión ha expirado. Por favor, inicia sesión nuevamente.';
          localStorage.removeItem('token');
          localStorage.removeItem('userRole');
          window.location.href = '/customer/login';
          break;
        case 404:
          errorMessage = error.error?.message || 'Carrito no encontrado';
          break;
        case 409:
          errorMessage = error.error?.message || 'Conflicto al procesar la orden';
          break;
        case 410:
          errorMessage = error.error?.message || 'El carrito ha expirado';
          break;
        case 500:
          errorMessage = error.error?.message || 'Error interno del servidor';
          break;
        default:
          errorMessage = error.error?.message || `Error del servidor: ${error.status}`;
      }
    }
    
    console.error('OrderService Error:', errorMessage, error);
    return throwError(() => new Error(errorMessage));
  }
}
